tensorpipe/channel/cuda_basic/context.cc

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <tensorpipe/channel/cuda_basic/context.h>
+
+#include <algorithm>
+#include <list>
+
+#include <tensorpipe/channel/cuda_basic/channel.h>
+#include <tensorpipe/channel/cuda_basic/context_impl.h>
+#include <tensorpipe/channel/error.h>
+#include <tensorpipe/channel/helpers.h>
+#include <tensorpipe/common/callback.h>
+#include <tensorpipe/common/cuda_loop.h>
+#include <tensorpipe/common/defs.h>
+#include <tensorpipe/common/error.h>
+#include <tensorpipe/common/error_macros.h>
+
+namespace tensorpipe {
+namespace channel {
+namespace cuda_basic {
+
+class Context::Impl : public Context::PrivateIface,
+                      public std::enable_shared_from_this<Context::Impl> {
+ public:
+  Impl(std::shared_ptr<CpuContext>);
+
+  const std::string& domainDescriptor() const;
+
+  std::shared_ptr<channel::CudaChannel> createChannel(
+      std::shared_ptr<transport::Connection>,
+      Endpoint);
+
+  void setId(std::string id);
+
+  ClosingEmitter& getClosingEmitter() override;
+
+  void close();
+
+  void join();
+
+  ~Impl() override = default;
+
+ private:
+  std::string domainDescriptor_;
+  std::atomic<bool> closed_{false};
+  std::atomic<bool> joined_{false};
+  ClosingEmitter closingEmitter_;
+
+  std::shared_ptr<CpuContext> cpuContext_;
+  std::shared_ptr<CudaLoop> cudaLoop_;
+  // An identifier for the context, composed of the identifier for the context,
+  // combined with the channel's name. It will only be used for logging and
+  // debugging purposes.
+  std::string id_{"N/A"};
+
+  // Sequence numbers for the channels created by this context, used to create
+  // their identifiers based off this context's identifier. They will only be
+  // used for logging and debugging.
+  std::atomic<uint64_t> channelCounter_{0};
+};
+
+Context::Context(std::shared_ptr<CpuContext> cpuContext)
+    : impl_(std::make_shared<Impl>(std::move(cpuContext))) {}
+
+Context::Impl::Impl(std::shared_ptr<CpuContext> cpuContext)
+    : domainDescriptor_("any"),
```

Comment:
Shouldn't this be the domain descriptor of the cpuChannel? In fact, we might not even have to store it, we can just forward any call to domainDescriptor() to the cpuChannel.